tests/cases/conformance/types/typeParameters/typeParameterLists/typeParameterConstModifiers.ts

```diff
@@ -0,0 +1,60 @@
+// @strict: true
+
+declare function f1<const T>(x: T): T;
+
+const x11 = f1('a');
+const x12 = f1(['a', ['b', 'c']]);
+const x13 = f1({ a: 1, b: "c", d: ["e", 2, true, { f: "g" }] });
+
+declare function f2<const T, U>(x: T | undefined): T;
+
+const x21 = f2('a');
+const x22 = f2(['a', ['b', 'c']]);
+const x23 = f2({ a: 1, b: "c", d: ["e", 2, true, { f: "g" }] });
+
+declare function f3<const T>(x: T): T[];
+
+const x31 = f3("hello");
+const x32 = f3("hello");
+
+declare function f4<const T>(obj: [T, T]): T;
+
+const x41 = f4([[1, 'x'], [2, 'y']]);
+const x42 = f4([{ a: 1, b: 'x' }, { a: 2, b: 'y' }]);
+
+declare function f5<const T>(obj: { x: T, y: T }): T;
+
+const x51 = f5({ x: [1, 'x'], y: [2, 'y'] });
+const x52 = f5({ x: { a: 1, b: 'x' }, y: { a: 2, b: 'y' } });
+
+declare function f6<const T extends readonly unknown[]>(...args: T): T;
+
+const x61 = f6(1, 'b', { a: 1, b: 'x' });
+
+class C1<const T> {
+    constructor(x: T) {}
+    foo<const U>(x: U) { return x; }
+}
+
+const c71 = new C1({ a: 1, b: "c", d: ["e", 2, true, { f: "g" }] });
+const c72 = c71.foo(['a', ['b', 'c']]);
+
+interface I1<const T> { x: T }  // Error
+
+type T1<const T> = T;  // Error
+
+// Corrected repro from #51745
+
+type Obj = { a: { b: { c: "123" } } };
+
+type GetPath<T, P> =
+    P extends readonly [] ? T :
+    P extends readonly [infer A extends keyof T, ...infer Rest] ? GetPath<T[A], Rest> :
+    never;
+
+function set<T, const P extends readonly string[]>(obj: T, path: P, value: GetPath<T, P>) {}
+
+declare let obj: Obj;
+declare let value: "123";
+
+set(obj, ['a', 'b', 'c'], value);

```

Comment:
What I'm wondering about in my comment [here](https://github.com/microsoft/TypeScript/pull/51865#issuecomment-1348513002) could potentially be used to validate the second argument itself. Right now we need to rely on the third one being computed to `never` based on the invalid const path in the second one.

Reply:
You could use the `const` type parameter as a type argument in the constraint to implement further validation. For example:

```ts
type StateKeys<T> = T extends { states?: object } ? keyof T['states'] & string : never;

type StateConfig<S extends string = string> = {
    initial?: S,
    states?: Record<S, StateConfig>
}

declare function createMachine<const T extends StateConfig<StateKeys<T>>>(config: T): void
```

Reply:
This is a good technique - thank you for sharing it! Building on that I managed to also build a recursive version (as I need it): [TS playground](https://www.staging-typescript.org/play?ts=5.0.0-pr-51865-14#code/C4TwDgpgBAysCGwIGkIgM4B4AqA+KAvFNlBAB5IB2AJulAN5ToJLoD8AXFAPYBGAVhADGwKAF8AUFChsoAazTcAZsQDaAImaII6dQF0oAMibAATgEtKAcylQulCADcIpgNwTQkWCx058REnIqWgYTbXQuPkERcRk1TR9dA3snF3cPcGg4bQBhbkolcytMWxhSCggaOmYLa0ITWqsAGltsbNZy4LoAJWFuU2pMGstm71z8wqt-KF6hfsHbaWHrFulpdog8gqKhsxGmqHhKEFxbU+n6W0tzYHN4ABtOWHclxKfLtahVZChLeUUVG1EnouBstpMSp91j5UBgcBt0N89LhVlCEfDEkjTp9cC9xO4xOlqMJ7vBTNAlABXSgicz5KBCcnaACy8CEAAtLBBMHNKMxiJ1KiEwRMdhtYVg8Ad0XhcLgABS2XmTLjYCQASi4jm45mo6UZEBZbM5DnlH1+lBud3uXHU8HUqK0rC45uk8BdizW11uD1tSm43Adnte4Q9ULW-u4LrEqM+Mc98ek8bE6vcQA). I should clean this up to use a single type parameter or something - but that's beside the point here. 

Thank you ❤️